app.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd June 14, 2009
.Os "Mac OS X"
.Dt APP \&1 "CLIMac Reference Manual"
.Sh NAME                 \" Section Header - required - don't modify 
.Nm app
.Nd Find OS X applications by a range of criteria.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.\".Nm
.\".Ar app-name
.\".Nm
.\".Op Fl l
.\".Op Fl a \*(Ba Fl u Ar URL-scheme \*(Ba Fl t Ar UTI
.\".Nm
.\".Op Fl b Ar bundleID \*(Ba Fl w Ar windowID \*(Ba Fl f Ar file
.Nm
.Ar name
.Nm
.Op Fl l
.Op Fl a
.Nm
.Op Fl l
.Op Fl b \*(Ba Fl f \*(Ba Fl w \*(Ba Fl u \*(Ba t
.Ar argument
.Sh DESCRIPTION          \" Section Header - required - don't modify
.Pp
The
.Nm
utility prints the path at which application(s) are located. In the absence of any other flags or options, the single argument is interpreted as the name of an application, and the full path to its bundle is printed.
.Sh OPTIONS
.Pp
In each case, the absolute paths to the applications matching the criteria are printed. Applications are specified by:
.Bl -ohang
.It Fl l Ns / Ns Fl -all
.D1 All applications .
.It Fl a Ns / Ns Fl -active
.D1 The currently active Po frontmost Pc application .
.D1 With the Fl l No flag , all launched applications .
.It Fl u Ns / Ns Fl -scheme Aq Ar url-scheme
.D1 Application that opens URLs with the specified scheme .
.D1 With the Fl l No flag , all applications able to handle the specified URL scheme .
.It Fl t Ns / Ns Fl -type Aq Ar UTI
.D1 Application that opens files of the specified content-type Po Uniform Type Identifier Pc .
.D1 With the Fl l No flag , all applications able to open files of the specified content-type .
.Pp
.D1 Use Xr uti 1 to determine UTIs .
.It Fl b Ns / Ns Fl -bundle Aq Ar bundle-ID
.D1 Application that has the specified identifier . Pq The bundle ID is a unique identifier for an application that is defined in the Info.plist under the key Sq CFBundleIdentifier
.It Fl f Ns / Ns Fl -file Aq Ar file
.D1 Application that will open the specified file . I\&t is recommeded to use the Fl t No option in the general case , and the Fl f No option only when information about a specific file is needed .
.It Fl w Ns / Ns Fl -window Aq Ar window-ID
.D1 Application that owns the specified window .
.Pp
.D1 Use Xr wid 1 to determine window IDs .
.El
.Pp
Only the default application for the given criteria is printed unless the
.Fl l
flag is also given.
.\".Sh FILES                \" File used or created by the topic of the man page
.Sh EXAMPLES
.Bd -literal -offset indent
% app iTunes
/Applications/iTunes.app

% app -a
/Applications/Utilities/Terminal.app

% app -u https
/Applications/Safari.app

% app -l -u https
/Applications/Chromium.app
/Applications/Firefox.app
/Applications/Safari.app

% app -t public.tiff
/Applications/Preview.app

% app -f .zshrc
/Applications/TextEdit.app

% app -f /bin/ls
/Applications/Utilities/Terminal.app

% app -b com.apple.Dock
/System/Library/CoreServices/Dock.app
.Ed
.Sh DIAGNOSTICS
.Ex -std beep
.\".Sh COMPATIBILITY
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr setapp 1 ,
.Xr uti 1 ,
.Xr wid 1
.\" .Sh BUGS              \" Document known, unremedied bugs
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner
.Sh AUTHORS
.Pp
The
.Nm
utility was written by
.An Vacuous Virtuoso
.Aq http://lipidity.com Ns .

src/app.m
#import <getopt.h>

CG_EXTERN OSStatus _LSCopyAllApplicationURLs(CFArrayRef *);
CG_EXTERN CGError CGSConnectionGetPID(int cid, pid_t *outPID);
CG_EXTERN int CGSMainConnectionID(void);
CG_EXTERN CGError CGSGetWindowOwner(int cid, int wid, int *outOwner);

/* LSCopyDefault(Role)?Handler functions are only valid for explicitly set associations. */

@interface NSString (aBID) - (NSString *) cp4bid; @end
@implementation NSString (aBID)
- (NSString *) cp4bid; {
	NSString *retval = [[NSWorkspace sharedWorkspace] absolutePathForAppBundleWithIdentifier:self];
	if (retval == nil)
		warnx("No path for bundle ID '%s'", [self fileSystemRepresentation]);
	return retval;
}
@end

int main(int argc, const char * argv[]) {
	if (argc > 1) {
		NSAutoreleasePool *pool = [NSAutoreleasePool new];
		const struct option longopts[] = {
			/* list */
			{ "all", no_argument, NULL, 'l' },
			{ "active", no_argument, NULL, 'a' },
			/* find */
			{ "bundle", required_argument, NULL, 'b' },
			{ "window", required_argument, NULL, 'w' },
			{ "scheme", required_argument, NULL, 'u' },
			{ "type", required_argument, NULL, 't' },
			{ "file", required_argument, NULL, 'f' },
			{ NULL, 0, NULL, 0 }
		};
		int c;
		char action = 0;
		NSString *arg = NULL;
		BOOL listAll = NO;
		while ((c = getopt_long(argc, (char **)argv, "lab:w:u:t:f:", longopts, NULL)) != EOF) {
			switch (c) {
				case 'l':
					listAll ^= 1;
					break;
				case 'a':
				case 'b':
				case 'w':
				case 'u':
				case 't':
				case 'f':
					if (action != 0) {
						fputs("You may not specify more than one `-bwuta' option", stderr);
						goto usage;
					} else {
						action = c;
						if (action != 'a') {
							arg = (NSString *)CFStringCreateWithFileSystemRepresentation(NULL, optarg);
							if (arg == NULL)
								errc(1, EINVAL, "Argument for -%c", action);
						}
					}
					break;
				default:
					goto usage;
			}
		}
		if (argc != optind && argc != 2)
			goto usage;

		NSWorkspace *ws = [NSWorkspace sharedWorkspace];

		NSArray *all = NULL;
		NSString *str = nil;
		switch (action) {
			case 'a':
				if (listAll)
					all = [[[ws launchedApplications] valueForKey:@"NSApplicationPath"] copy];
				else
					str = [[[ws activeApplication] objectForKey:@"NSApplicationPath"] copy];
				break;
			case 'b':
				str = [[arg cp4bid] copy];
				break;
			case 'w': {
				NSScanner *scan = [NSScanner scannerWithString:arg];
				int wid = 0;
				int appConnection = 0;
				int appPid = 0;
				[scan scanInt:&wid];
				ProcessSerialNumber s = {0u, 0u};
				if (CGSGetWindowOwner(CGSMainConnectionID(), wid, &appConnection) == 0) {
					CGSConnectionGetPID(appConnection, &appPid);
					GetProcessForPID(appPid, &s);
					FSRef ref = {{0}};
					GetProcessBundleLocation(&s, &ref);
					CFURLRef appURL = CFURLCreateFromFSRef(NULL, &ref);
					if (appURL != NULL) {
						str = (NSString *)CFURLCopyFileSystemPath(appURL, kCFURLPOSIXPathStyle);
						CFRelease(appURL);
					}
				}
			}
				break;
			case 'u':
				if (listAll) {
					CFArrayRef bids = LSCopyAllHandlersForURLScheme((CFStringRef)arg);
					if (bids) {
						all = [[(NSArray *)bids valueForKey:@"cp4bid"] copy];
						CFRelease(bids);
					}
				} else {
					NSURL *inURL = [[NSURL alloc] initWithScheme:arg host:@"" path:@"/"];
					CFURLRef outURL = NULL;
					if ((LSGetApplicationForURL((CFURLRef)inURL, kLSRolesAll, NULL, (CFURLRef *)&outURL) == 0) && (outURL != NULL))
						str = (NSString *)CFURLCopyFileSystemPath(outURL, kCFURLPOSIXPathStyle);
					[inURL release];
#if 0
					CFStringRef bundle = LSCopyDefaultHandlerForURLScheme((CFStringRef)arg);
					if (bundle) {
						str = [[(NSString *)bundle cp4bid] copy];
						CFRelease(bundle);
					}
#endif
				}
				break;
			case 't':
				if (listAll) {
					CFArrayRef bids = LSCopyAllRoleHandlersForContentType((CFStringRef)arg, kLSRolesAll);
					if (bids) {
						all = [[(NSArray *)bids valueForKey:@"cp4bid"] copy];
						CFRelease(bids);
					}
				} else {
					CFStringRef bundle = LSCopyDefaultRoleHandlerForContentType((CFStringRef)arg, kLSRolesAll);
					if (bundle) {
						str = [[(NSString *)bundle cp4bid] copy];
						CFRelease(bundle);
					}
					if (str == NULL) {
						CFURLRef outURL = NULL;
						CFStringRef tag = UTTypeCopyPreferredTagWithClass((CFStringRef)arg, kUTTagClassMIMEType);
						if (tag) {
							LSCopyApplicationForMIMEType(tag, kLSRolesAll, &outURL);
							if (outURL != NULL) {
								str = (NSString *)CFURLCopyFileSystemPath(outURL, kCFURLPOSIXPathStyle);
								CFRelease(outURL);
							}
						}
						if (str == NULL) {
							tag = UTTypeCopyPreferredTagWithClass((CFStringRef)arg, kUTTagClassFilenameExtension);
							if (tag) {
								LSGetApplicationForInfo(kLSUnknownType, kLSUnknownCreator, tag, kLSRolesAll, NULL, &outURL);
								if (outURL != NULL) {
									str = (NSString *)CFURLCopyFileSystemPath(outURL, kCFURLPOSIXPathStyle);
								}
							}
						}
					}
				}
				break;
			case 'f': {
				NSString *absPath = ([(NSString *)arg isAbsolutePath]) ? (NSString *)arg : [[[NSFileManager defaultManager] currentDirectoryPath] stringByAppendingPathComponent:(NSString *)arg];
				NSString *appName = nil, *fileType = nil;
				[ws getInfoForFile:absPath application:&appName type:&fileType];
				if (appName)
					str = [appName copy];
			}
				break;
			default: {
				if (listAll) {
					CFArrayRef appURLs = NULL;
					if (_LSCopyAllApplicationURLs(&appURLs) == 0) {
						all = [[(NSArray *)appURLs valueForKey:@"path"] copy];
						CFRelease(appURLs);
					}
				} else {
					arg = (NSString *)CFStringCreateWithFileSystemRepresentation(NULL, argv[1]);
					if (arg) {
						str = [[ws fullPathForApplication:arg] copy];
					}
				}
				break;
			}
		}
		if (arg != NULL)
			CFRelease(arg);
		if (str != nil) {
			puts([str fileSystemRepresentation]);
			CFRelease(str);
		} else if (all != nil) {
			NSEnumerator *e = [all objectEnumerator];
			NSString *obj;
			while ((obj = [e nextObject]) != nil)
				if ([obj respondsToSelector:@selector(fileSystemRepresentation)])
					puts([obj fileSystemRepresentation]);
			CFRelease(all);
		} else {
			exit(1);
		}
		[pool release];
		return 0;
	} else {
usage:
		fprintf(stderr, "usage:  %s <name>\n", argv[0]);
		const char * const uses[] = {
			"-l", "-a", "-b <bundle-id>", "-w <window-id>", "-u <url-scheme>", "-t <UTI>", "-f <file>"
		};
		size_t j = 0;
		do {
			fprintf(stderr, "\t%s %s\n", argv[0], uses[j]);
			j += 1;
		} while (j < (sizeof(uses)/sizeof(uses[0])));
		return 1;
	}
}
